logstash-core/lib/logstash/agent.rb

```diff
@@ -10,80 +10,125 @@
 LogStash::Environment.load_locale!
 
 class LogStash::Agent
-  attr_reader :logger
+  attr_reader :logger, :state
 
-  def initialize(logger)
-    @logger = logger
+  def initialize(params)
+    @logger = params[:logger]
     @pipelines = {}
+    @pipeline_threads = {}
+    @state = clean_state
+    @config_loader = LogStash::Config::Loader.new(@logger, false)
+    @config_string = params[:config_string]
+    @config_path = params[:config_path]
+    @auto_reload = params[:auto_reload]
+    @pipeline_settings = params[:pipeline_settings]
+    @reload_interval = params[:reload_interval] || 5 # seconds
   end
 
-  def config_valid?(config_str)
-    begin
-      # There should be a better way to test this ideally
-      LogStash::Pipeline.new(config_str)
-    rescue Exception => e
-      e
+  def execute
+    @thread = Thread.current
+    @logger.info("starting agent", :state => @state)
+
+    if @auto_reload
+      Stud.interval(@reload_interval) do
+        break unless clean_state? || running_pipelines?
```

Comment:
`break unless clean_state? || running_pipelines?` takes a minute to understand.

`reload_state! if empty_config? || running_pipelines?` is much more comprehensible IMHO

Since this is duplicated in later lines it may be even better to refactor this to:

`reload_state! if config_needs_reload?`